### ISSUE TITLE
t.conch.ssh test failures with openssh-9.8

### ISSUE THREAD

**Issue by arkamar** [twisted/twisted#12273]
twisted-24.7.0rc1 tests fail with openssh-9.8. It affects twisted-24.3.0 as well. Everything works if I downgrade to openssl-9.7.

Test fails with following erros/failures:
```
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "/var/tmp/portage/dev-python/twisted-24.7.0_rc1/work/twisted-24.7.0rc1-python3_12/install/usr/lib/python3.12/site-packages/twisted/conch/test/test_cftp.py", line 1510, in check
    self.assertEqual(result[2], 0, result[1].decode("ascii"))
  File "/var/tmp/portage/dev-python/twisted-24.7.0_rc1/work/twisted-24.7.0rc1-python3_12/install/usr/lib/python3.12/site-packages/twisted/trial/_synctest.py", line 444, in assertEqual
    super().assertEqual(first, second, msg)
  File "/usr/lib/python3.12/unittest/case.py", line 885, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/lib/python3.12/unittest/case.py", line 878, in _baseAssertEqual
    raise self.failureException(msg)
twisted.trial.unittest.FailTest: 255 != 0 : Load key "dsa_test": invalid format
testuser@127.0.0.1: Permission denied (publickey).
Connection closed


twisted.conch.test.test_cftp.OurServerSftpClientTests.test_extendedAttributes
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "/var/tmp/portage/dev-python/twisted-24.7.0_rc1/work/twisted-24.7.0rc1-python3_12/install/usr/lib/python3.12/site-packages/twisted/trial/_synctest.py", line 444, in assertEqual
    super().assertEqual(first, second, msg)
  File "/usr/lib/python3.12/unittest/case.py", line 885, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/lib/python3.12/unittest/case.py", line 878, in _baseAssertEqual
    raise self.failureException(msg)
twisted.trial.unittest.FailTest: None != b'test\n'

twisted.conch.test.test_conch.OpenSSHClientForwardingTests.test_localToRemoteForwarding
twisted.conch.test.test_conch.OpenSSHClientForwardingTests.test_localToRemoteForwardingV6
twisted.conch.test.test_conch.OpenSSHClientForwardingTests.test_remoteToLocalForwarding
===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.conch.error.ConchError: ('exit code was not 0: 255 (Load key "dsa_test": invalid format\r\ntestuser@127.0.0.1: Permission denied (publickey).\r\n)', None)

twisted.conch.test.test_conch.OpenSSHClientForwardingTests.test_exec
twisted.conch.test.test_conch.OpenSSHKeyExchangeTests.test_DH_GROUP14
twisted.conch.test.test_conch.OpenSSHKeyExchangeTests.test_DH_GROUP_EXCHANGE_SHA1
twisted.conch.test.test_conch.OpenSSHKeyExchangeTests.test_ECDHSHA256
twisted.conch.test.test_conch.OpenSSHKeyExchangeTests.test_DH_GROUP_EXCHANGE_SHA256
twisted.conch.test.test_conch.OpenSSHKeyExchangeTests.test_ECDHSHA384
twisted.conch.test.test_conch.OpenSSHKeyExchangeTests.test_ECDHSHA521
===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
DelayedCalls: (set twisted.internet.base.DelayedCall.debug = True to debug)
<DelayedCall 0x7f4f8e243a10 [0.0847017765045166s] called=0 cancelled=0 ConchTestForwardingProcess._connect()>

twisted.conch.test.test_conch.OpenSSHClientForwardingTests.test_localToRemoteForwarding
===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
DelayedCalls: (set twisted.internet.base.DelayedCall.debug = True to debug)
<DelayedCall 0x7f67d39f0a70 [0.08317971229553223s] called=0 cancelled=0 ConchTestForwardingProcess._connect()>

twisted.conch.test.test_conch.OpenSSHClientForwardingTests.test_localToRemoteForwardingV6
===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
DelayedCalls: (set twisted.internet.base.DelayedCall.debug = True to debug)
<DelayedCall 0x7f2f338f7d10 [0.08639383316040039s] called=0 cancelled=0 ConchTestForwardingProcess._connect()>

twisted.conch.test.test_conch.OpenSSHClientForwardingTests.test_remoteToLocalForwarding
===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.conch.error.ConchError: ('exit code was not 0: 255 (Pseudo-terminal will not be allocated because stdin is not a terminal.\r\nLoad key "dsa_test": invalid format\r\ntestuser@127.0.0.1: Permission denied (publickey).\r\n)', None)

twisted.conch.test.test_conch.OpenSSHClientRekeyTests.test_clientRekey
-------------------------------------------------------------------------------
Ran 11451 tests in 52.982s

FAILED (skips=764, failures=4, errors=11, successes=10675)
```

**Testing environment**
 - Operating System and Version; paste the output of these commands:
   - on Linux, `uname -a ; cat /etc/lsb-release`
   ```
   Linux gentoo-6c68faed 6.8.0 #1 SMP PREEMPT_DYNAMIC Wed Mar 13 11:30:08 CET 2024 x86_64 AMD Ryzen 5 2400G GNU/Linux
   cat: /etc/lsb-release: No such file or directory
   ```
   It is Gentoo system in container
 - Twisted version 24.7.0rc1

**Comment by adiroiban** [comment]
Thanks Petr for the report.

>  It affects twisted-24.3.0 as well

If 24.3.0 is also affect, I will not consider this a release blocker.

---

We have to check why our automated tests are not executed with OpenSSH 9.8

**Comment by adiroiban** [comment]
This looks like an issue with newer OpenSSH which might no longer have support for DSA/DSS

Not really a bug in Twisted.

We run the tests only with the OpenSSH version available in  22.04.4.

That is the only version that is "officially" supported.


I have Ubuntu 24.04 and it comes with OpenSSH is 9.6


--------

I guess that at some point Ubuntu will have OpenSSH 9.8 ... so we need to prepare for it :)

What can be done:

* Update the CI to show the OpenSSH version as part of the logs. For ex run: `ssh -V` 
* Update the CI to run conch tests inside a Gentoo container to have access to a newer OpenSSH version.
* Update the tests to skip the DSA tests if OpenSSH does not support it.

------

**Comment by cjwatson** [comment]
This'll be a problem for Debian shortly too.

I suggest testing whether the output of `ssh -Q key` contains a line reading "ssh-dss".  If it does, then you can test with it.  If it doesn't, you should skip the relevant tests.  I assume you don't care about ancient versions of OpenSSH (before 6.3) that don't support `ssh -Q key` at all, but if you do then you could just assume that they support DSA keys.
